### ISSUE TITLE
Editorconfig should not apply styles to files located outside the root directory

### ISSUE THREAD

**Issue by Tyriar** [editorconfig/editorconfig-vscode#52]
I've found that editorconfig changes to tabs as specified in the project's `.editorconfig` when I edit my settings.json file which is located away from the root open folder in `~/.config`. It should instead not apply the editorconfig settings in this case.

**Comment by jedmao** [comment]
I'm not sure I 100% understand your directory structure, but wouldn't adding an `.editorconfig` file in your `~/.config` folder that looks like this work for you?

```ini
root = true
```
That will stop it from looking for any more configuration settings when you open up files within that folder.

**Comment by Tyriar** [comment]
The indentation of my key bindings and settings files is all messed up because the project's editor config file overrides the VS code default detection. I don't want editor config to touch files outside of the folder, it's out of the scope of it so it shouldn't do it.

**Comment by SamVerschueren** [comment]
Are you sure this is because of the EditorConfig plugin? I scanned the code real quick and it shouldn't do that, so if it is the case, it's definitely a bug.

I recorded a screen capture and tried editing and saving it. As you can see, I have `trim_trailing_whitespace` set to `true` and `insert_final_newline` as well. Neither both of those transformations are being applied.

![editorconfig](https://cloud.githubusercontent.com/assets/1913805/18739162/e4792b76-80a1-11e6-81f2-b60bb1589683.gif)

I know it might be a stupid question, but I have to ask :). You aren't editing your workspace defined settings, right? Probably not because key bindings are always global, but you never know.

Care to screen capture the behaviour? Also, what OS are you using?

**Comment by Tyriar** [comment]
I open my settings and it gets set to tabs: 4 despite the file being indented with spaces: 2

![image](https://cloud.githubusercontent.com/assets/2193314/18756245/0a9a5dc8-80a4-11e6-9639-927627938589.png)

I disabled editorconfig and it works fine

![image](https://cloud.githubusercontent.com/assets/2193314/18756293/392dadde-80a4-11e6-805e-021d106e2f4c.png)

It is definitely an issue with the extension.

**Comment by SamVerschueren** [comment]
Care to share your `.editorconfig` file? I tested this on my system and it does not use the `.editorconfig` `indent_size` setting. As you can see, it is set to `2` and `settings.json` uses the one defined in the settings (4).

<img width="969" alt="screen shot 2016-09-22 at 19 23 50" src="https://cloud.githubusercontent.com/assets/1913805/18758862/21fe8894-80fa-11e6-9094-f3ab9da96530.png">

**Comment by Tyriar** [comment]
https://github.com/Microsoft/vscode/blob/master/.editorconfig

**Comment by jedmao** [comment]
Well, that's interesting and confusing at the same time!

**Comment by Tyriar** [comment]
Does [this](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L108) apply tab config before the `.editorconfig` is determined to relevant [here](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L137)?

**Comment by jedmao** [comment]
I think not, because the `tab_size` wouldn't be equal to anything if it wasn't relevant; thus, that block of code would not be executed.

**Comment by SamVerschueren** [comment]
TL;DR: I have a solution, will create a PR.

Sorry for this long explanation.

Alright, I started debugging this. Please note that this is on my computer with my setup. I don't have any `.editorconfig` files defined in home directories or whatsoever. This is important as `EditorConfig` will automatically traverse all the way up to the root untill it finds a `.editorconfig` file.

Let's get started!

When I open the global settings, the [`DocumentWatcher._onDidOpenTextDocument`](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L94) kicks in. Because the config was not yet loaded, it will ask `EditorConfig` to [parse the path](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L106). This is expected as stated before, it could be possible that you have a `.editorconfig` in your home directory. In my case, that promises just returns an empty object as no `.editorconfig` file was found.

Then it calls the [applyEditorConfigToTextEditor](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L153) method. The [settings](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L163) for the document are empty, e.g. `{}`.

----------------------------

###### sidetrack
The main question we have to ask first, is that behaviour correct? I already opened an issue on the [editorconfig-core-js](https://github.com/editorconfig/editorconfig-core-js/issues/36).

The thing is, if `editorconfig.parse()` would've returned `undefined`, it would [exit](https://github.com/editorconfig/editorconfig-vscode/blob/dcec5955f3ddbc65a154f206473cd863c32cfd96/src/DocumentWatcher.ts#L165) early and we're all happy.

This made me wonder. We already had Paul Irish creating issue #44 for this behaviour and I proposed fix #38 which was rejected in favour of #43. The thing is that #38 would've fixed this issue while #43 only made sure the message didn't get displayed.

-----------------------------

So let's continue where I left. After that, the [Utils.fromEditorConfig](https://github.com/editorconfig/editorconfig-vscode/blob/master/src/Utils.ts#L15) are called and create an options object based on the `.editorconfig` file (which is `{}`) and the editor settings.

And this is what causes the issue. It will read your workspace settings BUT it *doesn't* take the `editor.detectIndentation` into account resulting in the hard fixed settings defined in `settings.json`.

Let me create a PR that solves this issue.

**Comment by Tyriar** [comment]
@SamVerschueren thanks for looking into this, it's been bugging me for some time :stuck_out_tongue:

**Comment by Tyriar** [comment]
Still doesn't seem to be fixed with 0.3.2

**Comment by SamVerschueren** [comment]
@Tyriar You are right! 

@jedmao It looks like the code wasn't compiled before releasing a new version.

Source: `DocumentWatchter.ts`

```ts
private _onConfigChanged() {
	const workspaceConfig = workspace.getConfiguration('editor');
	const detectIndentation = workspaceConfig.get<boolean>('detectIndentation');

	this._defaults = (detectIndentation) ? {} : {
		tabSize: workspaceConfig.get<string | number>('tabSize'),
		insertSpaces: workspaceConfig.get<string | boolean>('insertSpaces')
	};
}
```

Code in Extension: `DocumentWatcher.js`

```js
_onConfigChanged() {
	this._defaults = {
		tabSize: vscode_1.workspace.getConfiguration('editor')
			.get('tabSize'),
		insertSpaces: vscode_1.workspace.getConfiguration('editor')
			.get('insertSpaces')
	};
}
```

**Comment by jedmao** [comment]
Sorry about that. I don't think the `vsce` tool has a prepublish script; otherwise, this would be much more preventable. Should be fixed now in v0.3.3.

**Comment by SamVerschueren** [comment]
@bpasero @Tyriar Is it currently possible to add a pre-publish hook for vsce or not? Might be a nice addition. Just making sure before I add a new issue in that repo.

**Comment by Tyriar** [comment]
Verified fixed :+1: thanks guys
